src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeWriter.cs

```diff
@@ -0,0 +1,316 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.IO.Pipelines
+{
+    internal class StreamPipeWriter : PipeWriter
+    {
+        internal const int InitialSegmentPoolSize = 4; // 16K
+        internal const int MaxSegmentPoolSize = 256; // 1MB
+
+        private readonly int _minimumBufferSize;
+
+        private BufferSegment _head;
+        private BufferSegment _tail;
+        private Memory<byte> _tailMemory;
+        private int _tailBytesBuffered;
+        private int _bytesBuffered;
+
+        private MemoryPool<byte> _pool;
+
+        private CancellationTokenSource _internalTokenSource;
+        private bool _isCompleted;
+        private object _lockObject = new object();
+
+        private BufferSegmentStack _bufferSegmentPool;
+
+        private CancellationTokenSource InternalTokenSource
+        {
+            get
+            {
+                lock (_lockObject)
+                {
+                    if (_internalTokenSource == null)
+                    {
+                        _internalTokenSource = new CancellationTokenSource();
+                    }
+                    return _internalTokenSource;
+                }
+            }
+        }
+
+        public StreamPipeWriter(Stream writingStream, StreamPipeWriterOptions options)
+        {
+            InnerStream = writingStream ?? throw new ArgumentNullException(nameof(writingStream));
+
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            _minimumBufferSize = options.MinimumBufferSize;
+            _pool = options.Pool == MemoryPool<byte>.Shared ? null : options.Pool;
+            _bufferSegmentPool = new BufferSegmentStack(InitialSegmentPoolSize);
+        }
+
+        /// <summary>
+        /// Gets the inner stream that is being written to.
+        /// </summary>
+        public Stream InnerStream { get; }
+
+        /// <inheritdoc />
+        public override void Advance(int bytes)
+        {
+            if ((uint)bytes > (uint)_tailMemory.Length)
```

Comment:
This doesn't work if bytes is a negative number. Check bytes < 0 before this check.

Reply:
Nvrm I'm incorrect.